src/util/samplebuffer.h

```diff
@@ -34,79 +35,94 @@ namespace mixxx {
 //
 class SampleBuffer final {
   public:
-    SampleBuffer()
-        : m_data(nullptr),
-          m_size(0) {
+    constexpr SampleBuffer()
+            : m_data(nullptr),
+              m_size(0) {
+    }
+    explicit SampleBuffer(SINT size)
+            : m_data((size > 0) ? SampleUtil::alloc(size) : nullptr),
+              m_size((m_data != nullptr) ? size : 0) {
     }
-    explicit SampleBuffer(SINT size);
     SampleBuffer(SampleBuffer&) = delete;
     SampleBuffer(SampleBuffer&& that)
-        : m_data(that.m_data),
-          m_size(that.m_size) {
-        that.m_data = nullptr;
-        that.m_size = 0;
+            : m_data(std::exchange(that.m_data, nullptr)),
+              m_size(std::exchange(that.m_size, 0)) {
+    }
+    ~SampleBuffer() {
+        SampleUtil::free(m_data);
     }
-    virtual ~SampleBuffer() final;
 
     SampleBuffer& operator=(SampleBuffer& that) = delete;
     SampleBuffer& operator=(SampleBuffer&& that) {
-        swap(that);
+        SampleUtil::free(m_data);
+        m_data = std::exchange(that.m_data, nullptr);
+        m_size = std::exchange(that.m_size, 0);
         return *this;
     }
 
-    constexpr SINT size() const {
+    constexpr SINT size() const noexcept {
```

Comment:
I am confused about the use of noexept in this class. IMHO it has a little value in this case since the function will be inlined anyway. 
On the other hand you did not add noexcept to the move constructors and move assignment operator as explicit recommended in the "C++ Core Guidelines" and other places. In these cases it may help std:: containers to optimize the code.

We do not use noexcept everywhere is Mixxx rightnow, because Mixxx does not use exceptions. So this is a new topic for us, which is ortogonal to to the main topic of this PR. Introducing noexcept to Mixxx can be considered as unnecessary boilerplate on one hand that drew developer resources on the other hand it may help under certain conditions.  

In practice, many codebases do not use noexcept on trivial getter functions, and it's not considered a best practice to do so unless there are specific reasons. The Mixxx style guide references the Google style guild here which is also kept weak: "Specify noexcept when it is useful and correct." 
  
What do you think, is it "useful" here?

Reply:
I also made a remark on that above. I would not have added it, but it's not wrong here.

The main culprit is that it's not really related to the PR and therefore increases the diff size which is a non-issue here because @Swiftb0y already reviewed it.

You "requested changes" in your review. Do you think it's absolutely necessary to remove these before merge? Otherwise I'd like to spare @abique the additional work and just keep it.

Reply:
I have requested a change, because it looks wrong to add noexcept to all functions in the class but not to the move constructors. 

I have no experience with except but I have read a bit and discovered three established styles: 
* Use noexcept everywhere 
 ** when using exceptions, can be errorphrone, extra boilerplate to maintain. 
* Use noexcept where useful
 ** unclear what useful is, google style 
* Use noexcept only on move assignment operators an constructors
** to allow optimazion in the std lib 
* don't use noexcept if no exceptions are used. 

I like to know at least what is the reasoning for the current changes. @abique

Reply:
The sample buffer class is meant to be used in real-time context.
Having the noexcept keywords means both that it should not throw exception (for the dev writting code), and that it won't throw exception (for the compiler).

Indeed I didn't annotate the move ctor and assignment, it is an overlook.
Though the assignment one, calls a free function which wasn't annotated itself, so I wasn't 100% sure about that one, but yes it unlikely that the free() will be the one to throw. And yes, having those two annotated will be useful if you have vectors of audio buffers which you resize.

Reply:
I don't claim to be a killer C++ dev, who knows everything.
But if the standard C++ library use those, there is certainly a good reason.

Reply:
> But if the standard C++ library use those, there is certainly a good reason.

We are no library, so different reasons apply here. 

After reading ab it about noexecpt it should be a good middle ground to use noexcept for all move ctor and assignments and not clutter the code with the noexcept everywhere. 

In this case I don't mind about he extra noexcept in this PR, but please add it to move ctors and assignments

Reply:
Done.

Reply:
I have started a topic in Zulip to fix the decision in our style guide: 
https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/noexcept